utils/merkle_proof/src/lib.rs

```diff
@@ -1,129 +1,137 @@
+    }
+}
 
-        hashes
-    };
+fn concat(mut vec1: Vec<u8>, mut vec2: Vec<u8>) -> Vec<u8> {
+    vec1.append(&mut vec2);
+    return vec1;
+}
 
-    static ref ZERO_NODES: Vec<MerkleTree> = {
-        (0..=MAX_TREE_DEPTH).map(MerkleTree::Zero).collect()
-    };
+fn hash_and_concat(h1: H256, h2: H256) -> H256 {
+    H256::from_slice(&hash(&concat(
+        h1.as_bytes().to_vec(),
+        h2.as_bytes().to_vec(),
+    )))
 }
 
-#[derive(Debug)]
-pub enum MerkleTree {
-    Leaf(H256),
-    Node(H256, Box<Self>, Box<Self>),
-    Zero(usize),
+fn get_previous_power_of_two(x: usize) -> usize {
+    if x <= 2 {
+        return x;
+    } else {
+        return 2 * get_previous_power_of_two(x/2);
+    }
 }
 
-impl MerkleTree {
-    pub fn create(leaves: &[H256], depth: usize) -> Self {
-        use MerkleTree::*;
-        if leaves.is_empty() {
-             return Zero(depth); 
-        }
+fn get_next_power_of_two(x: usize) -> usize {
+    if x <= 2 {
+        return x;
+    } else {
+        return 2 * get_next_power_of_two((x+1)/2);
+    }    
+}
 
-        if depth == 0 {
-                assert_eq!(leaves.len(), 1);
-                Leaf(leaves[0])
-        } else {
-                let capacity = get_next_power_of_two(depth-1);
-                let (l_leaves, r_leaves) = if leaves.len() <= capacity { (leaves, EMPTY_SLICE) } else { leaves.split_at(capacity)};
-                let l_subtree = MerkleTree::create(l_leaves, depth - 1);
-                let r_subtree = MerkleTree::create(r_leaves, depth - 1);
-                let hash = hash_and_concat(l_subtree.hash(), r_subtree.hash());
-                Node(hash, Box::new(l_subtree), Box::new(r_subtree))
-        }
+fn concat_generalized_indices(indices: &[usize]) -> usize {
+    let mut o = 1usize;
+    for index in indices.iter() {
+        o = o * get_previous_power_of_two(*index) + (index - get_previous_power_of_two(*index));
     }
+    return o;
+}
 
-    pub fn hash(&self) -> H256 { 
-        match *self {
-            MerkleTree::Leaf(h) => h,
-            MerkleTree::Node(h, _, _) => h,
-            MerkleTree::Zero(depth) => ZERO_HASHES[depth],
-        }
-    }
+fn get_generalized_index_length(index: usize) -> usize {
+   return log_of!(index, 2., usize);
+}
 
-    pub fn left_and_right_branches(&self) -> Option<(&Self, &Self)> {
-        match *self {
-            MerkleTree::Leaf(_) | MerkleTree::Zero(0) => None,
-            MerkleTree::Node(_, ref l, ref r) => Some((l, r)),
-            MerkleTree::Zero(depth) => Some((&ZERO_NODES[depth - 1], &ZERO_NODES[depth - 1])),
-        }
-    }
+fn get_generalized_index_bit(index: usize, position: usize) -> bool {
+    // ((index >> position) & 0x01) == 1 lighthouse 
+    ((index >> position) & 0x01) > 0 //dokumentacija
+}
 
-    pub fn is_leaf(&self) -> bool {
-        match self {
-            MerkleTree::Leaf(_) => true,
-            _ => false,
-        }
-    }
+fn generalized_index_sibling(index: usize) -> usize {
+    return index^1;
```

Comment:
This is really the power of 1? Why not just return index then?

Reply:
This is XOR operator.